solr/contrib/ltr/src/java/org/apache/solr/ltr/LTRInterleavingRescorer.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.ltr;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TopDocs;
+import org.apache.solr.ltr.interleaving.Interleaving;
+import org.apache.solr.ltr.interleaving.InterleavingResult;
+import org.apache.solr.ltr.interleaving.TeamDraftInterleaving;
+
+/**
+ * Implements the rescoring logic. The top documents returned by solr with their
+ * original scores, will be processed by a {@link LTRScoringQuery} that will assign a
+ * new score to each document. The top documents will be resorted based on the
+ * new score.
+ * */
+public class LTRInterleavingRescorer extends LTRRescorer {
+  
+  LTRScoringQuery[] rerankingQueries;
+  Interleaving interleavingAlgorithm = new TeamDraftInterleaving();
```

Comment:
11.1/n If we go with 4/n then `LTRQParserPlugin.LTRQParser` could pass a `Interleaving interleavingAlgorithm` argument to the `LTRInterleavingQuery` constructor which could pass it to the `LTRInterleavingRescorer` constructor. For now `TeamDraftInterleaving` would be the only supported algorithm but in future other algorithms could then easily be added e.g. based on an additional `ltr` parameter. What do you think? An easy change to make now or something better left for later?

Reply:
I added this bit, I think is easy enough and not much invasive to prepare the structure now.

Adding new algorithm would be as easy as :
- fetching the param from the request
- add the algorithm name in the switch clause
- implement the new algorithm

Take a look if you like it